File: python/Figure3_4.py
# Author: Kenji Kashima
# Date  : 2025/09/01
# Note  : pip install control

import control
import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(1)

def create_noise_model(Ts):
    # --- 1. Noise Model time discretization ---
    # Continuous-time noise model F = 1/(s+0.3)
    s = control.TransferFunction.s
    F_c = 1 / (s + 0.3)
    F_ss = control.ss(F_c)
    F_d = control.c2d(F_ss, Ts, method='tustin')
    Aw, Bw, Cw = F_d.A, F_d.B, F_d.C

    # --- 2. Noise Model Normalization ---
    # Solve discrete Lyapunov equation: P = A * P * A^T + B * B^T
    Pw = control.dlyap(Aw, Bw @ Bw.T)
    variance_amplification = (Cw @ Pw @ Cw.T)[0, 0]  # Scalar variance
    Bw = Bw / np.sqrt(variance_amplification)

    return Aw, Bw, Cw

def figure3_4a(Aw, Bw, Cw):
    figsize = config.global_config(type=1)
    dsys = control.ss(Aw, Bw, Cw, 0, dt=True)
    mag, _, omega = control.frequency_response(dsys, np.arange(0.001, np.pi, 0.001))
    plt.figure(figsize=figsize)
    plt.semilogx(omega, mag, linewidth=1.0, color='blue', label='frequency weight')
    plt.plot([0.001,0.01,0.6,np.pi], [8,8,0,0], linewidth=2, linestyle='--', color='red', label='prior information')
    plt.xlabel(r'$\varpi$')
    plt.grid()
    plt.legend()
    plt.tight_layout()
    plt.savefig("./Figure3_4a.pdf")
    plt.show()

def figure3_4b( Aw, Bw, Cw, k_bar ):
    x_bar_dim = Aw.shape[0]
    v = np.random.randn(k_bar+1) # white noise
    x_bar = np.zeros((x_bar_dim, k_bar + 1)) # state of weighting filter
    y = np.zeros(k_bar + 1) # colored noise
    
    for k in range(k_bar):
        x_bar[:, k+1] = Aw @ x_bar[:, k] + Bw.ravel() * v[k]
        y[k+1] = (Cw @ x_bar[:, k+1])[0]

    figsize = config.global_config(type=1)
    plt.figure(figsize=figsize)
    plt.xlabel(r'$k$')
    plt.xlim(0,200)
    plt.ylim(-2,2)
    plt.stairs(v[:201], label='white')
    plt.stairs(y[:201], linewidth=1.0, label='colored')
    plt.legend(loc='upper right')
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure3_4b.pdf")
    plt.show()

# Parameters for noise model
Ts = 0.1
Aw, Bw, Cw = create_noise_model(Ts)
# nw = Aw.shape[0]

if __name__ == '__main__':
    figure3_4a(Aw, Bw, Cw)
    figure3_4b(Aw, Bw, Cw, k_bar=200)

File: matlab/Figure3_4.m
% Author: Kenji Kashima
% Date  : 2025/09/01
% Note  : requires Control System Toolbox

clear;close all; rng(1); % random seed

%% Create noise model
%% --- 1. Noise Model Normalization ---
% Continuous-time noise model F = 1/(s+0.3)
s = tf('s');
F_c = 1 / (s + 0.3);
F_ss = ss(F_c);
Ts = 0.1;
F_d = c2d(F_ss, Ts, 'tustin');
Aw = F_d.A;
Bw = F_d.B;
Cw = F_d.C;

%% --- 2. Noise Model Normalization ---
% Solve discrete Lyapunov equation: Pw = Aw * Pw * Aw' + Bw * Bw'
Pw = dlyap(Aw, Bw*Bw');
variance_amplification = Cw * Pw * Cw';   % scalar variance
Bw = Bw / sqrt(variance_amplification(1,1));


% Figure 3.4(a)
% 離散時間システム
dsys = ss(Aw, Bw, Cw, 0, Ts);

% 周波数応答（0.001 ~ π）
omega = (0.001:0.001:pi).';           % 列ベクトルにしておく
H = freqresp(dsys, omega / Ts);
mag = abs(squeeze(H)); 
figure('Name','Figure 3.4(a)'); hold on; grid on;

semilogx(omega, mag, ...
    'LineWidth', 1.0, 'Color', 'b', ...
    'DisplayName', 'frequency weight');

semilogx([0.001, 0.01, 0.6, pi], [8, 8, 0, 0], ...
    'LineWidth', 2, 'LineStyle', '--', 'Color', 'r', ...
    'DisplayName', 'prior information');

set(gca, 'XScale', 'log');   % 念のため強制
xlim([1e-3, pi]);
xlabel('$\varpi$','Fontsize',18,'Interpreter','latex');
ylabel('Magnitude','Fontsize',18,'Interpreter','latex');
legend('Fontsize',16,'Interpreter','latex');
movegui('northwest');

% Figure 3.4(b)
k_bar = 200;
x_bar_dim = size(Aw,1);
v = randn(1, k_bar+1);             % white noise
x_bar = zeros(x_bar_dim, k_bar+1); % state of weighting filter
y = zeros(1, k_bar+1);             % colored noise

for k = 1:k_bar
    x_bar(:,k+1) = Aw * x_bar(:,k) + Bw * v(k);

    y(k+1) = Cw * x_bar(:,k+1);
end

figure('Name','Figure 3.4(b)'); hold on; grid on;
stairs(v)
stairs(y,'linewidth',1)
xlim([0,200])
ylim([-2,2])
xlabel('$k$','Fontsize',18,'Interpreter','latex')
legend('white','colored','Fontsize',16)
movegui('northeast')
